File: Csharp/Algorithm/dpShortPath.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Algorithm
{
    public class DPShortPath
    {
        public class Edge
        {
            public string src, dest;

            public int layer;

            public float weight;
        }

        public class Graph
        {
            public List<Edge> edges = new List<Edge>();

            public int layerCout;
        }

        public static float Fn(Graph graph, string src, int layer)
        {
            if (layer == graph.layerCout)
            {
                return 0;
            }

            var edges = graph.edges.FindAll(val => val.layer == layer & & val.src == src);

            return edges.Min(val => (val.weight + Fn(graph, val.dest, layer + 1)));
        }

        public static void Init()
        {
            var graph = new Graph
            {
                layerCout = 3
            };

            //A->B
            graph.edges.Add(new Edge()
            {
                src = "A",
                dest = "B",
                layer = 0,
                weight = 4
            });

            //A->C
            graph.edges.Add(new Edge()
            {
                src = "A",
                dest = "C",
                layer = 0,
                weight = 2
            });

            //B->E
            graph.edges.Add(new Edge()
            {
                src = "B",
                dest = "E",
                layer = 1,
                weight = 3
            });

            //B->D
            graph.edges.Add(new Edge()
            {
                src = "B",
                dest = "D",
                layer = 1,
                weight = 7
            });

            //C->D
            graph.edges.Add(new Edge()
            {
                src = "C",
                dest = "D",
                layer = 1,
                weight = 5
            });

            //E->F
            graph.edges.Add(new Edge()
            {
                src = "E",
                dest = "F",
                layer = 2,
                weight = 4
            });

            //D->F
            graph.edges.Add(new Edge()
            {
                src = "D",
                dest = "F",
                layer = 2,
                weight = 3
            });

            Console.WriteLine(Fn(graph, "A", 0));
        }
    }
}

File: Cpp/dpShortPath.cpp
#include <iostream>

using namespace std;

struct Edge
{
    string src, dest;
    
    float weight;
};

struct Graph
{
    // V-> Number of vertices, E-> Number of edges
    int V, E;

    // graph is represented as an array of edges.
    struct Edge *edge;
};

struct Graph *createGraph(int V, int E)
{
    struct Graph *graph = new Graph;
    graph->V = V;
    graph->E = E;
    graph->edge = new Edge[E];
    return graph;
}

int main()
{
    int V = 6;
    int E = 7;

    struct Graph *graph = createGraph(V, E);

    graph->edge[0].src = "A";
    graph->edge[0].dest = "B";
    graph->edge[0].weight = 4;

    graph->edge[1].src = "A";
    graph->edge[1].dest = "C";
    graph->edge[1].weight = 2;

    graph->edge[2].src = "B";
    graph->edge[2].dest = "E";
    graph->edge[2].weight = 2;

    graph->edge[3].src = "B";
    graph->edge[3].dest = "D";
    graph->edge[3].weight = 7;

    graph->edge[4].src = "C";
    graph->edge[4].dest = "D";
    graph->edge[4].weight = 5;

    graph->edge[5].src = "E";
    graph->edge[5].dest = "F";
    graph->edge[5].weight = 4;

    graph->edge[6].src = "D";
    graph->edge[6].dest = "F";
    graph->edge[6].weight = 3;

}